_source/_posts/2023-03-20-spa-auth-tokens.md

```diff
@@ -0,0 +1,228 @@
+
+If you haven't already done so, open the SPA code in your favorite IDE and navigate to the `config.js` or `app.config.ts` file, depending on your framework. You'll see the OIDC config and a property for `scopes` where you will add `offline_access`. Your OIDC config looks something like this:
+
+```js
+{
+    clientId: CLIENT_ID,
+    issuer: ISSUER,
+    redirectUri: window.location.origin + '/login/callback',
+    scopes: ['openid', 'profile', 'email', 'offline_access']
+  }
+```
+
+Now try rerunning the SPA. If you inspect the network request, you'll see the refresh token too. You can see the tokens by looking at the contents of your Local Storage too.
+
+## Use auth tokens in SPAs
+
+Now that we have these tokens in our SPA, what do we do with them? We should let a trusted OIDC library, such as the Okta SDKs we're using, handle all the token requests and refresh them for us. Behind the scenes the OIDC library is hard at work exchanging tokens.
+
+{% img blog/spa-auth-tokens/refresh-token.jpg alt:"Refresh token call to authorization server and returning new tokens" width:"800" %}{: .center-image }
+
+We're primarily interested in the contents of the ID token and using the Access token.
+
+### How to use the Access Token
+The Access token gives us access to resources. That means we use Access tokens in outgoing API requests that our application needs, like for Sunny to get their next rawring lesson.
+
+{% img blog/spa-auth-tokens/access-token.jpg alt:"Sunny starting up a new lesson makes a call to the Rawr site's /lesson endpoint passing in the Access token" width:"800" %}{: .center-image }
+
+You'll send the Access token as a Bearer token in the HTTP call's Authorization header. So in the case of Sunny, the outgoing HTTP call looks like this.
+
+```http
+GET /lessons
+Authorization: Bearer base64encoded_access_token
+```
+
+You'll want to ensure you're not adding the Access token to calls outside your system by maintaining an allowlist of origins that should include the token.
+
+### How to use the ID Token
+Since the ID Token contains user identity information, you can use this token to start populating your user store, and for supporting identifiers you need immediately, such as their name.
+
+{% img blog/spa-auth-tokens/id-token.jpg alt:"Sunny looking at his profile in the Rawr app where they see their name, dinosaur family, and their avatar" width:"800" %}{: .center-image }
+
+Once you have the Access token, you can make a request to the OIDC Discovery document's User Info endpoint, which may contain more information about the user than what is available in your ID token. You probably will also need to add user information and user settings from API calls within your system since the SPAs we write have complex user information.
```

Comment:
```suggestion
Once you have the access token, you can make a request to the OIDC Discovery document's user info endpoint, which may contain more information about the user than what is available in your ID token. You probably will also need to add user information and user settings from API calls within your system since the SPAs we write have complex user information.
```